org-code-codebuilderapp/app/src/main/java/org/code/codebuilderapp/CodeBuildController.java

```diff
@@ -0,0 +1,33 @@
+package org.code.codebuilderapp;
+
+import java.security.Principal;
+import org.springframework.messaging.handler.annotation.MessageMapping;
+import org.springframework.messaging.simp.annotation.SendToUser;
+import org.springframework.stereotype.Controller;
+
+/**
+ * Accepts requests to the /execute channel to compile and run user code.
+ * Directs output from the user's program to the user on the /topic/output
+ * channel.
+ */
+@Controller
+public class CodeBuildController {
+
+  private final CompileRunService compileRunService;
+
+  CodeBuildController(CompileRunService compileRunService) {
+    this.compileRunService = compileRunService;
+  }
```

Comment:
(minor) In case you didn't already know, I think there's some Spring annotation that will automatically inject this dependency.

Reply:
Thank you! Keep the minor tips coming! I'm so new to Spring that a lot of this still feels like magic :)

Reply:
Heh, at some point, I knew more about Spring autowiring than I wanted to and it's definitely black magic.  Which reminds me that the annotation that I used was `@Autowired` (although there may be newer, more specific annotations).  IIRC, for a class to be a target for autowiring, it also has to be annotated with @Service, @Controller, etc.  (Let me know if you have questions and I can look at the docs to give you more precise answers.)

Reply:
Actually, a little help here would be nice. The docs around autowiring aren't the most descriptive. I'm having trouble understanding how it works or what the use case is for it. (most of the docs say something along the lines of "the annotation @Autowired allows you to autowire a constructor" 🙄)